crates/dex-swap-router/src/rpc.rs

```diff
@@ -0,0 +1,442 @@
+use crate::*;
+use dex_stable::traits::{StablePair, StableSwapMode};
+use sp_runtime::{FixedPointNumber, FixedPointOperand, FixedU128};
+use sp_std::vec;
+
+const MAX_HOPS: u32 = 4;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Amount<Balance, CurrencyId> {
+    value: Balance,
+    currency: CurrencyId,
+}
+
+impl<Balance, CurrencyId> Amount<Balance, CurrencyId> {
+    pub fn new(value: Balance, currency: CurrencyId) -> Self {
+        Self { value, currency }
+    }
+}
+
+pub trait DexQuery<Balance, CurrencyId> {
+    type Path;
+    fn path_of(&self, input: &CurrencyId) -> Self::Path;
+    fn contains(&self, currency: &CurrencyId) -> bool;
+    fn get_output_amount(&self, amount_in: &Amount<Balance, CurrencyId>) -> Option<Amount<Balance, CurrencyId>>;
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum DexTradingPair<T: Config> {
+    Stable(StablePair<T::StablePoolId, T::CurrencyId>),
+    Normal(T::CurrencyId, T::CurrencyId),
+}
+
+impl<T: Config> DexQuery<T::Balance, T::CurrencyId> for DexTradingPair<T> {
+    type Path = Route<T::StablePoolId, T::CurrencyId>;
+
+    fn path_of(&self, input: &T::CurrencyId) -> Self::Path {
+        match self {
+            Self::Stable(stable_pair) => Self::Path::Stable(stable_pair.clone().path_of(input.clone())),
+            Self::Normal(token0, token1) => Self::Path::Normal(vec![
+                input.clone(),
+                if input == token0 { token1 } else { token0 }.clone(),
+            ]),
+        }
+    }
+
+    fn contains(&self, currency: &T::CurrencyId) -> bool {
+        match self {
+            Self::Stable(stable_pair) => &stable_pair.token0 == currency || &stable_pair.token1 == currency,
+            Self::Normal(token0, token1) => token0 == currency || token1 == currency,
+        }
+    }
+
+    fn get_output_amount(
```

Comment:
This function is not very well tested. I think only `rpc_find_best_trade` hits this code code at all